contracts/src/v0.8/ccip/tokenAdminRegistry/TokenPoolFactory/TokenPoolFactory.sol

```diff
@@ -0,0 +1,289 @@
+    _setTokenPoolInTokenAdminRegistry(token, pool);
+
+    // Begin the 2 step ownership transfer of the newly deployed token to the msg.sender
+    IOwnable(token).transferOwnership(msg.sender);
+
+    return (token, pool);
+  }
+
+  /// @notice Deploys a token pool with an existing ERC20 token
+  /// @dev Since the token already exists, this contract is not the owner and therefore cannot configure the
+  /// token pool in the token admin registry in the same transaction. The user must invoke the calls to the
+  /// tokenAdminRegistry manually
+  /// @dev since the token already exists, the owner must grant the mint and burn roles to the pool manually
+  /// @param token The address of the existing token to be used in the token pool
+  /// @param remoteTokenPools An array of remote token pools info to be used in the pool's applyChainUpdates function
+  /// @param tokenPoolInitCode The creation code for the token pool
+  /// @param salt The salt to be used in the create2 deployment of the token pool
+  /// @return poolAddress The address of the token pool that was deployed
+  function deployTokenPoolWithExistingToken(
+    address token,
+    RemoteTokenPoolInfo[] calldata remoteTokenPools,
+    bytes calldata tokenPoolInitCode,
+    bytes32 salt,
+    PoolType poolType
+  ) external returns (address poolAddress) {
+    // Ensure a unique deployment between senders even if the same input parameter is used to prevent
+    // DOS/Frontrunning attacks
+    salt = keccak256(abi.encodePacked(salt, msg.sender));
+
+    // create the token pool and return the address
+    return _createTokenPool(token, remoteTokenPools, tokenPoolInitCode, salt, poolType);
+  }
+
+  // ================================================================
+  // |                Pool Deployment/Configuration                  |
+  // ================================================================
+
+  /// @notice Deploys a token pool with the given token information and remote token pools
+  /// @param token The token to be used in the token pool
+  /// @param remoteTokenPools An array of remote token pools info to be used in the pool's applyChainUpdates function
+  /// @param tokenPoolInitCode The creation code for the token pool
+  /// @param salt The salt to be used in the create2 deployment of the token pool
+  /// @return poolAddress The address of the token pool that was deployed
+  function _createTokenPool(
+    address token,
+    RemoteTokenPoolInfo[] calldata remoteTokenPools,
+    bytes calldata tokenPoolInitCode,
+    bytes32 salt,
+    PoolType poolType
+  ) private returns (address) {
+    // Create an array of chain updates to apply to the token pool
+    TokenPool.ChainUpdate[] memory chainUpdates = new TokenPool.ChainUpdate[](remoteTokenPools.length);
+
+    RemoteTokenPoolInfo memory remoteTokenPool;
+    for (uint256 i = 0; i < remoteTokenPools.length; ++i) {
```

Comment:
nit: I think it is cheaper to read the length from storage once and then cache it ex

```solidity
numTokenPools = remoteTokenPools.length;
for (uint256 i = 0; i < numTokenPools; ++i) {...}
```